### PR TITLE
Add gossip Marshaller interface

### PR DESCRIPTION
**Pull request by joshua-kim**
## Why this should be merged

Required to unblock the X-Chain's integration for the sdk gossip

## How this works

Refactors marshal and unmarshal to be on its own interface instead of attached to the `Gossipable` interface

## How this was tested

UTs pass

### CODE REVIEW

File: network/p2p/gossip/gossipable.go [hunk 1]
```diff
@@ -8,8 +8,12 @@ import "github.com/ava-labs/avalanchego/ids"
...
 type Gossipable interface {
 	GetID() ids.ID
-	Marshal() ([]byte, error)
-	Unmarshal(bytes []byte) error
+}
+
+// Marshaller creates concrete Gossipable types from bytes
+type Marshaller[T Gossipable] interface {
+	GossipMarshal(T) ([]byte, error)
+	GossipUnmarshal([]byte) (T, error)
```
Comment (StephenButtolph) [inline network/p2p/gossip/gossipable.go:16]: I feel like this reads better:
```suggestion
	MarshalGossip(T) ([]byte, error)
	UnmarshalGossip([]byte) (T, error)
```

File: network/p2p/gossip/gossip_test.go [hunk 2]
```diff
@@ -10,7 +10,6 @@ import (
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
-
```
Comment (StephenButtolph) [inline network/p2p/gossip/gossip_test.go:13]: 👀 (nit reduce diff)

File: network/p2p/gossip/handler.go [hunk 3]
```diff
@@ -121,7 +124,7 @@ func (h Handler[_, U]) AppRequest(_ context.Context, _ ids.NodeID, _ time.Time,
 	return proto.Marshal(response)
 }
 
-func (h Handler[T, U]) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) {
+func (h Handler[T]) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) {
```
Comment (StephenButtolph) [inline network/p2p/gossip/handler.go:127]:
```suggestion
func (h Handler[_]) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) {
```

File: network/p2p/gossip/gossipable.go [hunk 4]
```diff
@@ -8,8 +8,12 @@ import "github.com/ava-labs/avalanchego/ids"
 // Gossipable is an item that can be gossiped across the network
 type Gossipable interface {
 	GetID() ids.ID
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply network/p2p/gossip/gossipable.go]: Renamed

File: network/p2p/gossip/gossipable.go [hunk 5]
```diff
@@ -7,9 +7,13 @@ import "github.com/ava-labs/avalanchego/ids"
 
 // Gossipable is an item that can be gossiped across the network
 type Gossipable interface {
-	GetID() ids.ID
-	Marshal() ([]byte, error)
-	Unmarshal(bytes []byte) error
+	GetGossipID() ids.ID
```
Comment (StephenButtolph) [inline network/p2p/gossip/gossipable.go:10]: I don't think golang generally uses getters (the `Get` prefix).
```suggestion
	GossipID() ids.ID
```

File: network/p2p/gossip/test_gossip.go [hunk 6]
```diff
@@ -10,26 +10,29 @@ import (
...
 }
 
-func (t *testTx) Unmarshal(bytes []byte) error {
-	t.id = ids.ID{}
-	copy(t.id[:], bytes)
-	return nil
+func (testMarshaller) UnmarshalGossip(bytes []byte) (*testTx, error) {
+	return &testTx{
+		id: ids.ID(bytes),
+	}, nil
```
Comment (StephenButtolph) [inline network/p2p/gossip/test_gossip.go:35]: This is supper nitty - but I think we should be using `ids.ToID` here. `ids.ID` will panic if `bytes` isn't the correct length.

Comment (joshua-kim) [reply network/p2p/gossip/test_gossip.go:35]: Hmm I don't think that's nitty ^_^

**Review by StephenButtolph — APPROVED** [review]

**Review by dhrubabasu — APPROVED** [review]

**Pull request merged**